[LBSE] Add RenderSVGResourceMasker

rwlbuis opened the pull request:
#### 08ff1e4f09f8c06c8ffae86dd8e1bb4feed8abe5
<pre>
[LBSE] Add RenderSVGResourceMasker
<a href="https://bugs.webkit.org/show_bug.cgi?id=264049">https://bugs.webkit.org/show_bug.cgi?id=264049</a>

Reviewed by Nikolas Zimmermann.

Implement the second SVG resource in LBSE: masking (RenderSVGResourceMasker).
Mask content invalidations/changes are supported.

The resource takes cyclic references into account.

Covered by existing tests (dozens of progressions, see TestExpectations).

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-intro-01-f-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-mask-01-b-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-root-position-masking-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/empty-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-mask-and-object-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-mask-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-changes-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-child-changes-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-colorspace-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-excessive-malloc-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-inside-defs-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-invalidation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-with-all-units-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-with-default-value-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/transformedMaskFails-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/visibility-override-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/filters/filter-clip-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/foreignObject/mask-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/text-mask-update-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink.html: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-shrink.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgMaskerResourceFromStyle const):
(WebCore::RenderLayerModelObject::paintSVGClippingMask const):
(WebCore::RenderLayerModelObject::paintSVGMask const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourceContainer const):
(WebCore::RenderObject::isRenderSVGResourceMasker const):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp: Added.
(WebCore::RenderSVGResourceMasker::RenderSVGResourceMasker):
(WebCore::createImageBuffer):
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceMaskerInlines.h: Added.
(WebCore::RenderSVGResourceMasker::maskElement const):
(WebCore::RenderSVGResourceMasker::maskUnits const):
(WebCore::RenderSVGResourceMasker::maskContentUnits const):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::paintSVGClippingMask): Deleted.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::createElementRenderer):
(WebCore::SVGMaskElement::calculateMaskContentRepaintRect):
* Source/WebCore/svg/SVGMaskElement.h:

Canonical link: <a href="https://commits.webkit.org/271153@main">https://commits.webkit.org/271153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e388f9d3bf5bca80e4597a778047f6ce0e819735

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27502 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6144 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29730 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25177 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3539 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24966 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27767 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4930 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23613 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4293 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4457 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24610 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30366 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25035 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30591 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2615 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/24345 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6617 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4865 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
```diff
@@ -133,8 +133,7 @@ void RenderSVGContainer::paint(PaintInfo& paintInfo, const LayoutPoint& paintOff
     if (paintInfo.phase == PaintPhase::Mask) {
-        // FIXME: [LBSE] Upstream SVGRenderSupport changes
-        // SVGRenderSupport::paintSVGMask(*this, paintInfo, adjustedPaintOffset);
+        SVGRenderSupport::paintSVGMask(*this, paintInfo, adjustedPaintOffset);
```
I think it's time to move them into a shared place, that is not SVGRenderSupport, but REnderLayerModelObject.h. There is also a section in that header, which already contains LBSE specific functions shared across RenderSVG* renderers. Let's move paintSVGClipMask and paintSVGMask there :-)

rwlbuis replied:
You are right, that way we can also make it non-static. Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,151 @@
+
+RenderSVGResourceMasker::~RenderSVGResourceMasker() = default;
+
+bool RenderSVGResourceMasker::applyResource(RenderElement&, const RenderStyle&, GraphicsContext*&, OptionSet<RenderSVGResourceMode>)
```
Soon this is obsolete. We can land this before the resource invalidation rework.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,151 @@
+
+FloatRect RenderSVGResourceMasker::resourceBoundingBox(const RenderObject& object, RepaintRectCalculation)
+{
+    FloatRect targetBoundingBox = object.objectBoundingBox();
```
auto?

rwlbuis replied:
Sure, fixed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,151 @@
+{
+    FloatRect targetBoundingBox = object.objectBoundingBox();
+
+    // Resource was not layouted yet. Give back the boundingBox of the object.
```
Not sure how relevant that is in LBSE.

rwlbuis replied:
Good point ,should it be an ASSERT instead or just remove the check?

nikolaszimmermann replied:
I'd say try the assert. Not sure if that's already possible before the LBSE rework patch landed (regarding invalidation).

rwlbuis replied:
Sure, done!

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,151 @@
+        maskRect = contentTransform.mapRect(maskRect);
+    }
+
+    FloatRect maskBoundaries = SVGLengthContext::resolveRectangle<SVGMaskElement>(&maskElement, maskElement.maskUnits(), targetBoundingBox);
```
auto?

rwlbuis replied:
Fixed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,151 @@
+
+    auto& maskElement = this->maskElement();
+
+    FloatRect maskRect = object.strokeBoundingBox();
```
auto?

rwlbuis replied:
Fixed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -678,4 +680,27 @@ void SVGRenderSupport::paintSVGClippingMask(const RenderLayerModelObject& render
+    if (!paintInfo.shouldPaintWithinRoot(renderer) || paintInfo.phase != PaintPhase::Mask || context.paintingDisabled())
+        return;
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
```
Evil boiler plate stuff, will get better with my patch to rework the SVG resource lookup.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
```diff
@@ -299,4 +299,32 @@ bool SVGRenderingContext::bufferForeground(RefPtr<ImageBuffer>& imageBuffer)
     return true;
 }
 
+RefPtr<ImageBuffer> SVGRenderingContext::createImageBuffer(const FloatRect& targetRect, const AffineTransform& absoluteTransform, const DestinationColorSpace& colorSpace, RenderingMode, const GraphicsContext* context)
```
Do we really want to add stuff to SVGRenderingContext? I am trying to make SVGRenderSupport / SVGRenderingContext legacy-only.

rwlbuis replied:
Sure, removed.

nikolaszimmermann reviewed (commented):
Great first stab.

nikolaszimmermann commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -86,6 +86,7 @@
 #include "SVGElementTypeHelpers.h"
 #include "SVGImage.h"
 #include "SVGLengthContext.h"
+#include "SVGRenderStyle.h"
```
Really needed?

rwlbuis replied:
Looks like not needed, removed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -558,6 +586,34 @@ void RenderLayerModelObject::repaintOrRelayoutAfterSVGTransformChange()
     repaintRendererOrClientsOfReferencedSVGResources();
 }
+
+void RenderLayerModelObject::paintSVGClippingMask(PaintInfo& paintInfo) const
```
The whole method needs a `#if ENABLE(LAYER_BASED_SVG_ENGINE)' guard.

rwlbuis replied:
It actually is already included in one.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -558,6 +586,34 @@ void RenderLayerModelObject::repaintOrRelayoutAfterSVGTransformChange()
+void RenderLayerModelObject::paintSVGMask(PaintInfo& paintInfo, const LayoutPoint& adjustedPaintOffset) const
+{
+    auto& context = paintInfo.context();
+    if (!paintInfo.shouldPaintWithinRoot(*this) || paintInfo.phase != PaintPhase::Mask || context.paintingDisabled())
```
IIRC we only call this for `PaintPhase::Mask`, no? Then you can add the same assertion as in `paintSVGClippingMask` just for `PaintPhase::Mask`.

rwlbuis replied:
Added the ASSERT.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.h:
```diff
@@ -95,6 +96,9 @@ class RenderLayerModelObject : public RenderElement {
     RenderSVGResourceClipper* svgClipperResourceFromStyle() const;
+    RenderSVGResourceMasker* svgMaskerResourceFromStyle() const;
+    void paintSVGClippingMask(PaintInfo&) const;
+    void paintSVGMask(PaintInfo&, const LayoutPoint&) const;
```
The necessity of 'adjustedPaintOffset' (the 2nd parameter) or the absence of it in `paintSVGClippingMask` should be revisited / clarified. IMHO one should spell out the parameter name here for clarity - if it doesn't violate style rules.

rwlbuis replied:
I added the name, indeed it can't be deduced from the parameter type.

nikolaszimmermann commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -469,12 +469,13 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
-    bool isRenderSVGResourceContainer() const { return type() == Type::SVGResourceClipper; }
+    bool isRenderSVGResourceContainer() const { return type() == Type::SVGResourceClipper || type() == Type::SVGResourceMasker; }
     bool isRenderSVGResourceFilter() const { return type() == Type::SVGResourceFilter; }
     bool isLegacyRenderSVGResourceMarker() const { return type() == Type::SVGResourceMarker; }
```
We should decide on a one, unique solution. Clipper has LegacySVGResourceClipper / SVGResourceClipper. For marker we only have SVGResourceMarker at present. I'd go for explicit LegacySVGResourceMasker / SVGResourceMasker.

rwlbuis replied:
You are right, confusing to be inconsistent, I added LegacySVGResourceMasker.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
```diff
@@ -127,14 +127,13 @@ void RenderSVGContainer::paint(PaintInfo& paintInfo, const LayoutPoint& paintOff
 
     if (paintInfo.phase == PaintPhase::ClippingMask) {
-        SVGRenderSupport::paintSVGClippingMask(*this, paintInfo);
+        paintSVGClippingMask(paintInfo);
```
Much better :-)

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,179 @@
+    FloatSize scale;
+    FloatSize clampedSize = ImageBuffer::clampedSize(paintRect.size(), scale);
+
+#if USE(DIRECT2D)
```
We still have DIRECT2D support?

rwlbuis replied:
No, seems removed, fixed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,179 @@
+    AffineTransform transform;
+    transform.scale(scale).translate(-paintRect.location()).multiply(absoluteTransform);
+
+    GraphicsContext& imageContext = imageBuffer->context();
```
s/GC/auto/

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,179 @@
+        return nullptr;
+
+    AffineTransform transform;
+    transform.scale(scale).translate(-paintRect.location()).multiply(absoluteTransform);
```
IIRC we have a generic function that can be used to create image buffers. Maybe I'm wrong. Let's check :-)

rwlbuis replied:
I tried quickly and saw a regression in svg/custom (did not try other directories). Since this could be some subtle code difference we agreed to leave this for later.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
```diff
@@ -0,0 +1,179 @@
+
+FloatRect RenderSVGResourceMasker::resourceBoundingBox(const RenderObject& object, RepaintRectCalculation repaintRectCalculation)
+{
+    ASSERT(!selfNeedsLayout());
```
I think this should go - and be in sync with how clipper does things.

rwlbuis replied:
Indeed that assertion is not relevant anymore, removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -236,7 +236,8 @@ std::unique_ptr<SVGResources> SVGResources::buildCachedResources(const RenderEle
                 treeScope.addPendingSVGResource(id, element);
 #else
-            treeScope.addPendingSVGResource(id, element);
+            else
```
This is no longer needed, LBSE doesn't use SVGResources anymore.

rwlbuis replied:
Good point, removed.

nikolaszimmermann commented on Source/WebCore/svg/SVGMaskElement.cpp:
```diff
@@ -130,7 +131,26 @@ void SVGMaskElement::childrenChanged(const ChildChange& change)
+FloatRect SVGMaskElement::calculateMaskContentRepaintRect(RepaintRectCalculation repaintRectCalculation)
+{
+    FloatRect maskRepaintRect;
+    for (Node* childNode = firstChild(); childNode; childNode = childNode->nextSibling()) {
```
auto in a few places.

rwlbuis replied:
Fixed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -558,6 +586,31 @@ void RenderLayerModelObject::repaintOrRelayoutAfterSVGTransformChange()
+{
+    ASSERT(paintInfo.phase == PaintPhase::Mask);
+    auto& context = paintInfo.context();
+    if (!paintInfo.shouldPaintWithinRoot(*this) || paintInfo.phase != PaintPhase::Mask || context.paintingDisabled())
```
The ::Mask condition in the if can go away.

rwlbuis replied:
Wel spotted, removed.

nikolaszimmermann reviewed (commented):
Looks great already.

nikolaszimmermann reviewed (approved):
Allright, let's continue when this is in.

rwlbuis committed:
[LBSE] Add RenderSVGResourceMasker
https://bugs.webkit.org/show_bug.cgi?id=264049

Reviewed by Nikolas Zimmermann.

Implement the second SVG resource in LBSE: masking (RenderSVGResourceMasker).
Mask content invalidations/changes are supported.

The resource takes cyclic references into account.

Covered by existing tests (dozens of progressions, see TestExpectations).

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-intro-01-f-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/W3C-SVG-1.1/masking-mask-01-b-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/absolute-root-position-masking-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/empty-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-mask-and-object-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/js-late-mask-creation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-changes-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-child-changes-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-colorspace-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-excessive-malloc-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-inside-defs-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-invalidation-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-with-all-units-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/mask-with-default-value-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/recursive-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/transformedMaskFails-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/visibility-override-mask-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/filters/filter-clip-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/foreignObject/mask-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/repaint/text-mask-update-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-shrink.html: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed.html: Added.
* LayoutTests/svg/repaint/mask-object-bounding-box.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-shrink-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-shrink.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed.html: Added.
* LayoutTests/svg/repaint/mask-user-space-on-use.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::svgMaskerResourceFromStyle const):
(WebCore::RenderLayerModelObject::paintSVGClippingMask const):
(WebCore::RenderLayerModelObject::paintSVGMask const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderSVGResourceContainer const):
(WebCore::RenderObject::isRenderSVGResourceMasker const):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp: Added.
(WebCore::RenderSVGResourceMasker::RenderSVGResourceMasker):
(WebCore::createImageBuffer):
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceMaskerInlines.h: Added.
(WebCore::RenderSVGResourceMasker::maskElement const):
(WebCore::RenderSVGResourceMasker::maskUnits const):
(WebCore::RenderSVGResourceMasker::maskContentUnits const):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::paintSVGClippingMask): Deleted.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::createElementRenderer):
(WebCore::SVGMaskElement::calculateMaskContentRepaintRect):
* Source/WebCore/svg/SVGMaskElement.h:

Canonical link: https://commits.webkit.org/271153@main
